guide/src/trait_bounds.md

```diff
@@ -0,0 +1,509 @@
+However, we can write a second wrapper around these functions to call them directly.
+This wrapper will also perform the type conversions between Python and Rust.
+
+```rust
+#[pymethods]
+impl UserModel {
+
+    pub fn set_variables_in_rust(&mut self, var: Vec<f64>) -> PyResult<()> {
+        println!("Set variables from Python calling Rust");
+        self.set_variables(&var);
+        Ok(())
+    }
+
+    pub fn get_results_in_rust(&mut self) -> PyResult<Vec<f64>> {
+        println!("Get results from Python calling Rust");
+        let results = self.get_results();
+        let gil = Python::acquire_gil();
+        let py = gil.python();
+        let py_results = results.into_py(py);
+        Ok(py_results)
+    }
+
+    pub fn compute_in_rust(&mut self) -> PyResult<()> {
+        println!("Compute from Python calling Rust");
+        self.compute();
+        Ok(())
+    }
+}
+```
+This wrapper handles the type conversion between the PyO3 requirements and the trait.
+In order to meet PyO3 requirements, this wrapper must:
+- return an object of type `PyResult`
+- use only values, not references in the method signatures
+
+Let's run the file python file:
+
+```python
+class Model:
+    def set_variables(self, inputs):
+        self.inputs = inputs
+    def compute(self):
+        self.results = [elt**2 - 3 for elt in self.inputs]
+    def get_results(self):
+        return self.results
+
+if __name__=="__main__":
+  import trait_exposure
+
+  myModel = Model()
+  my_rust_model = trait_exposure.UserModel(myModel)
+  my_rust_model.set_variables_in_rust([2.0])
+  print("Print value from Python: ", myModel.inputs)
+  my_rust_model.compute_in_rust()
+  print("Print value from Python: ", my_rust_model.get_results_in_rust())
+
+```
+
+This outputs:
+
+```block
+Set variables from Python calling Rust
+Set variables from Rust calling Python
+Print value from Python:  [2.0]
+Compute from Python calling Rust
+Compute from Rust calling Python
+Get results from Python calling Rust
+Get results from Rust calling Python
+Print value from Python:  [1.0]
+```
+
+We have now successfully exposed a Rust model that implements the `Model` trait to Python!
+
+We will now expose the `solve` function, but before, let's talk about types errors.
+
+## Type errors in Python
+
+What happens if you have type errors when using Python and how can you improve the error messages?
+
+
+### Wrong types in Python function arguments
+
+Let's assume in the first case that you will use in your Python file `my_rust_model.set_variables_in_rust(2.0)` instead of `my_rust_model.set_variables_in_rust([2.0])`.
+
+The Rust signature expects a vector, which corresponds to a list in Python.
+What happens if instead of a vector, we pass a single value ?
+
+At the execution of Python, we get :
+
+```block
+File "main.py", line 15, in <module>
+   my_rust_model.set_variables_in_rust(2)
+TypeError
+```
+
+It is a type error and Python points to it, so it's easy to identify and solve.
```

Comment:
One day I hope to add detail to these `TypeError`s!

Reply:
That would be an improvement! However, I think it is already easy to debug as Python points to the incorrect call.